### ISSUE TITLE
Analysis tool for class/sty file dependencies.

### ISSUE THREAD

**Issue by jbs1** [jbs1/tract2#21]
_migrated from Trac, where originally posted by **kohlhase** on 15-Oct-2006 6:42am_

This is done, thanks to Elchin and friends

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 15-Oct-2006 6:42am_

We need a tool that deterines the dependencies in the style/class files (see also bug 1340).

> The LaTeX class/style files have dependencies in themselves: there may be
> \input, \usepackage, \RequirePackage, and \LoadClass macros in the TeX sources
> (see the LaTeX documentation on what I may have missed). 

We need to have a Perl tool that determines these, so that they can be used in the LaTeXML bindings and the build system (see bugs 1339 and 1341)

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 15-Oct-2006 7:21am_

ooops wrong dependency

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 24-Oct-2006 9:38am_

As an alternative to building a tool, we could also make the files by hand. But there are quite a few of them, so as a starter, we could just grep for 

\LoadClass, and for every xxx.cls that has a \LoadClass that is not article.cls, we could change the 

RequirePackage('article'); 

in xxx.cls to the appropriate value. 

furthermore, we can add 

RequirePackage('yyy');

to xxx.ltxml for all occurrences of 

\input{yyy.sty}
\usepakcage{yyy}
\RequirePackage{yyy}

in xxx.cls

see arXMLiv/trunk/sty/bourbaki.ltxml for an example. I think that this is all we need to do for bourbaki.sty

That should eliminate a lot of errors. 

I am not sure that we can automate this, but it should only be a couple of hours work to do this by hand. 

And it would be nice, if we had this done before we start the next big run.

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 24-Oct-2006 10:01am_

I think a good start would be to manually do all the amsart-based classes and packages, since Ioan has already done a lot of work on the ams* packages.

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **ionita** on 30-Oct-2006 12:37am_

(In reply to comment #2)
> \LoadClass, and for every xxx.cls that has a \LoadClass that is not
> article.cls, we could change the 
> 
> RequirePackage('article'); 
> 
> in xxx.cls to the appropriate value. 

I found it difficult to detect the appropriate value that we should change in \RequirePackege, or in other words, how do we know the files in which amsart should be introduced (there has to be a reference in the .cls file?)?

And after we make changes -- is running a make on the sandbox and then seeing whether the classes don't have dependencies anymore -- a way of verifying that we did the right modifications?

Andrei

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 7-Nov-2006 7:51am_

(In reply to comment #4)
> (In reply to comment #2)
> > \LoadClass, and for every xxx.cls that has a \LoadClass that is not
> > article.cls, we could change the 
> > 
> > RequirePackage('article'); 
> > 
> > in xxx.cls to the appropriate value. 
> 
> I found it difficult to detect the appropriate value that we should change in
> \RequirePackege, or in other words, how do we know the files in which amsart
> should be introduced (there has to be a reference in the .cls file?)?

yes, the \LoadClass and \RequirePackage are these references!
> 
> And after we make changes -- is running a make on the sandbox and then seeing
> whether the classes don't have dependencies anymore -- a way of verifying that
> we did the right modifications?

Yes, that is exactly what should be done. 

Michael 
> Andrei
>

**Comment by jbs1** [comment]
_migrated from Trac, where originally posted by **kohlhase** on 7-Nov-2006 7:56am_

OK, now I know how these dependencies should be integrated into the build process. 

I have made a file arXMLiv/trun/sty/Makefile.stydeps that is included by the paper  Makefiles.  We can specify the dependencies there. 
For each foo.ltxml file that has a dependency (i.e. contains a RequirePackage('bar');) we must list the bar.ltxml in the rule for foo.ltxml after the colon.

For instance, if foo.ltxml contains

RequirePackage('bar');
RequirePacakge('baz');

then we have to have the line 

foo.ltxml: bar.ltxml baz.ltxml

in Makefile.stydeps (see the documentation in the file).

So it seems that this should be easy to implement in a little Perl skript.
